Propagate through calls more selectively to avoid false positives

mlevesquedion opened the pull request:
Specifically, avoid propagating in the absence of knowledge about a function's behavior. As a mitigation, summaries for a large number of standard library functions are added and used, which should give reasonable coverage in the intraprocedural case.

## Context

The impetus for this PR came from issues involving our assumption that, when the function we are analyzing calls another function with a tainted argument, we should assume that the other arguments and the return value are all tainted. This assumption is unworkable in practice, as it leads to frequent false positives, essentially turning the analyzer into a "was a function called with a tainted value" detector. In such cases, the burden of analysis is shifted onto a human, who has to analyze the entire call tree below the "offending" call.

In future work, we will invest in developing an actual interprocedural analysis, probably based on some kind of pointer analysis.

With the changes in this PR, there are no longer any false positives on `kubernetes/kubernetes`.

## Process

I tried to gather standard library functions that I think could reasonably be involved in a propagation. I did not look at all functions. More summaries can easily be added later if desired. Also, 

Packages I covered:
```
fmt
errors
strings
bytes
io
io/ioutil
bufio
context
strconv
encoding/json
encoding/base64
sync
text
log
path
path/filepath
```

## Tips for reviewers

This PR is meant as a single atomic operation that changes the way we propagate through calls. As such, it contains the removal of some tests which are no longer relevant/appropriate. It also contains the introduction of tests for new code.

The core changes are made in `propagation/propagation.go` and the new `propagation/stdlib.go` file. The bulk of the lines added come from `propagation/summaries.go`, which contains the function summaries (these are just data, not code).

- [x] Running against a large codebase such as [Kubernetes](https://github.com/kubernetes/kubernetes) does not error out. (See [DEVELOPING.md](https://github.com/google/go-flow-levee/blob/master/DEVELOPING.md) for instructions on how to do that.)
- (N/A) [ ] Appropriate changes to README are included in PR

PurelyApplied commented on internal/pkg/propagation/propagation.go:
```diff
@@ -254,37 +252,10 @@ func (prop *Propagation) taintCall(call *ssa.Call, maxInstrReached map[*ssa.Basi
-	prop.taintReferrers(call, maxInstrReached, lastBlockVisited)
-	for _, a := range call.Call.Args {
-		prop.taintCallArg(a, maxInstrReached, lastBlockVisited)
+	if callee := call.Call.StaticCallee(); callee != nil {
```
Can a `call` be both `StdlibCall` and `StdlibInterfaceCall`?

mlevesquedion replied:
Unless I've made a mistake, no. None of the functions in `funcSummaries` match the name/signature pairs in `interfaceFuncSummaries`.

PurelyApplied replied:
An explicit `if / else` or an early `return` in the `if` block would be warranted, then.  From a program-flow perspective, you're indicating that a thing could be both.

PurelyApplied commented on internal/pkg/propagation/summaries.go:
```diff
@@ -0,0 +1,761 @@
+	// There is a 1-to-1 mapping between the bits and the function's
+	// parameters, with the least significant bit corresponding to the
+	// first (0th) argument.
+	ifTainted int64
```
This feels like a premature optimization.  I don't see why this is a bitwise `int64` but `taintedArgs` / `taintedRets` are `[]int`.

If you're worried about performance, a `map[int]struct{args, rets []int}` mapping input tainted-arg to output tainted-whatever would still be well usable and reasonably readable.  I don't see this struct being a performance or memory bottleneck, though.

------

Coming back to this now that I've read everything else, it's not that bad.  I think it's more the mismatch that's bothering me.  I like the utility of `fromFirstArgToFirstRet`, and I feel like that could be generalized into something like

```golang
const (
  first int64 = 1 << 1
  second
  third
  fourth
)

var funcSummaries = map[string]summary{
  // func Errorf(format string, a ...interface{}) error {
  "fmt.Errorf": {
    ifTainted:   first | second,
    taintedRets: first,
  },
  // ...
}
```

Either way, I think I'd like there to be consistency between the "input slice" and "output slices".

mlevesquedion replied:
(I don't anticipate performance to be an issue here. FWIW, and I should probably have mentioned this somewhere, AFAICT these changes have no performance impact.) I typically keep bitwise manipulation as a last resort, but in this case I think it leads to the most readable code, specifically because it makes it very easy to test whether the intersection is empty or not.

I agree that it would be nice to have consistency between the inputs and outputs, but I think it makes more sense to have separate data representations because they are used in different ways. The bitset makes sense when looking for taint because we're interested in the intersection. The slices make sense when looking for where to propagate taint, because we can just iterate over them.

I really like your `first, second, third, ...` idea for the bits. I think this is a great improvement to readability. I have implemented it.

PurelyApplied replied:
Glad you adopted the `first | second` approach.  I still think it would read a bit better if the other fields were consistent with it, but I'm not going to bikeshed on it.  Marking this thread resolved.

PurelyApplied commented on internal/pkg/propagation/stdlib.go:
```diff
@@ -0,0 +1,154 @@
+		return
+	}
+
+	if call.Call.Signature().Results().Len() == 1 {
+		if len(summ.taintedRets) > 0 {
+			prop.taintReferrers(call, maxInstrReached, lastBlockVisited)
+		}
+		return
+	}
+
+	indexToExtract := map[int]*ssa.Extract{}
+	for _, r := range *call.Referrers() {
+		e := r.(*ssa.Extract)
+		indexToExtract[e.Index] = e
+	}
+	for i := range summ.taintedRets {
+		prop.taint(indexToExtract[i], maxInstrReached, lastBlockVisited, true)
+	}
```
I think I follow this only because I am deeply familiar with `ssa`.  Comments throughout this method could be useful.  Then again, "deeply familiar with `ssa`" is rather the intended audience at this depth.

mlevesquedion replied:
Added some high-level comments.

PurelyApplied commented on internal/pkg/propagation/stdlib.go:
```diff
@@ -0,0 +1,154 @@
+// is known (i.e. the function has a summary).
+func (prop *Propagation) taintStdlibCall(call *ssa.Call, maxInstrReached map[*ssa.BasicBlock]int, lastBlockVisited *ssa.BasicBlock) {
+	summ, ok := funcSummaries[funcName(call)]
+	if !ok {
+		return
+	}
```
How exhaustive is the current summary set?  Would it be better to explicitly enumerate the "no taint propagated" methods and indicate error here if an unemumerate function is found?

mlevesquedion replied:
For the set of packages I covered, I have good confidence that I have covered the relevant functions. For the packages I chose not to cover, I don't know if they may contain relevant functions. I could cover the remaining packages and add any new summaries in a future PR?

PurelyApplied replied:
I was more asking if, for the packages covered, if the functions included in the summaries constituted more than, say, 90% of the functions that were present in those packages.  I don't expect there to be much churn in the stdlib without us noticing, but it could be a good defensive measure to make sure we're not skipping over some functions without deliberately having excluded them and/or end up with a function signature that should be matching a summary but isn't due to a bug.

That investigation could also be opened as an Issue and rolled out as another PR, though.

mlevesquedion replied:
Hmm. From memory, I think for a good number of packages it was a significant fraction of the exported functions, but I would need to double check. Also we may want to monitor Go releases for new functions. Opened #293.

PurelyApplied commented on internal/pkg/propagation/stdlib.go:
```diff
@@ -0,0 +1,154 @@
+//   WriteTo(w Writer) (n int64, err error)
+// The result is:
+//   (Writer)(int64,error)
+func sigTypeString(sig *types.Signature) string {
```
`[Testing]` This feels potentially brittle and like something we shouldn't have to implement ourselves.  Then again, we're only using it to conform built-ins to the summary-table we've made, so it's probably fine within that scope.

mlevesquedion replied:
That was my initial thought. `Signature` does implement `Stringer`, but it includes the names of things in addition to their types, which doesn't work for our use case.

I agree that this needs dedicated testing, I will add such tests shortly.

PurelyApplied commented on internal/pkg/propagation/summaries.go:
```diff
@@ -0,0 +1,761 @@
+		ifTainted:   0b10,
+		taintedArgs: []int{0},
+	},
+	// func Sscan(str string, a ...interface{}) (n int, err error) {
```
`[Testing]` I need to review how scanning is used.

PurelyApplied commented on internal/pkg/propagation/summaries.go:
```diff
@@ -0,0 +1,761 @@
+	},
+	// func Map(mapping func(rune) rune, s string) string {
+	"strings.Map": {
+		ifTainted:   0b10,
```
Do we have any concept of a tainted `func`?  I suppose we could contrive one by injecting a source through a closure as a freevar, but that's probably well out of scope for us.  (Also applicable to many `func` below.)

mlevesquedion replied:
This is a good question. I agree that this is out of scope, without pointer analysis I don't think we could model this with good enough precision.

PurelyApplied commented on internal/pkg/propagation/summaries.go:
```diff
@@ -0,0 +1,761 @@
+	"strings.TrimPrefix": fromFirstArgToFirstRet,
+	// func TrimSuffix(s, suffix string) string {
+	"strings.TrimSuffix": fromFirstArgToFirstRet,
+	// func Replace(s, old, new string, n int) string {
```
This is a curious case.  I could see someone using `Replace` to specifically purge a secret, e.g. `Replace(problematic, secret, "REDACTED", -1)`.  Probably not an issue until someone hits it, but we might want to make a note.

Ditto `ReplaceAll`.

mlevesquedion replied:
Yeah, some of these are a bit weird. When in doubt I leaned towards including a function rather than excluding it.

Since we check for sanitizers before propagating, I think this should be fine, i.e. if a user really wants to use string replacements to perform sanitization, they could configure `Replace*` to be sanitizers.

RE `we might want to make a note`: would you like me to include the above reasoning in a comment somewhere?

PurelyApplied replied:
Good points.  And yeah, I don't think there's really a good place to make a note of it that is somewhere we would remember to reference.  The Note will be if someone opens an Issue for it, then.

PurelyApplied commented on internal/pkg/propagation/summaries.go:
```diff
@@ -0,0 +1,761 @@
+	"bufio.NewReaderSize": fromFirstArgToFirstRet,
+	// func NewReader(rd io.Reader) *Reader {
+	"bufio.NewReader": fromFirstArgToFirstRet,
+	// TODO: test
```
`[Testing]` This comment highlights this `// TODO`.

mlevesquedion replied:
Oops. Please don't tell anyone how I live.

PurelyApplied commented on internal/pkg/propagation/summaries.go:
```diff
@@ -0,0 +1,761 @@
+// such as Write or Read, that could be called statically (i.e. a call to
+// a concrete method whose signature matches an interface method) or dynamically
+// (i.e. a call to an interface method on an interface value).
+var interfaceFuncSummaries = map[funcKey]summary{
```
My understanding of the current code is that `arg0` is the receiver for all `ifTainted` uses below.  That would be a good comment here.

mlevesquedion replied:
Done.

PurelyApplied commented on internal/pkg/propagation/summaries.go:
```diff
@@ -0,0 +1,761 @@
+		taintedRets: []int{0},
+	},
+	// type fmt.GoStringer interface {
+	//	GoString() string
+	// }
+	{"String", "()(string)"}: {
```
Mismatch.

mlevesquedion replied:
Good catch. Done.

PurelyApplied commented on internal/pkg/propagation/summaries.go:
```diff
@@ -0,0 +1,761 @@
+		ifTainted:   0b1,
+		taintedRets: []int{0},
+	},
+}
```
`[Testing]` Review of these summaries was end of day while my eyes twitched from allergies.  I should do another pass for better confidence.

PurelyApplied reviewed (changes requested):
This is good.  I'm very glad to hear that restricting the propagation set leaves us with no false-positives in Kubernetes.

This is a significant departure from our existing model.  While it might seem like I'm needling with some of my comments, it's really more that I'm trying to make sure we're placing our feet carefully.  That doesn't always carry through in text, so I thought it bore explicitly stating.  Overall, I'm very happy with this PR.

I've read the core code changes, but not the tests yet.  I want to take a slow and careful look and the change in our testing contract.  Some of the comments I make in the code are marked with `[Testing]` to reflect that the answer to those concerns either are already or can be resolved by adequate testing, and I will flesh out those comments early next week.  Most of them are more TODO reminders for myself mid-review than anything else.

------

With the introduction of the two new files, the `propagation` package is at roughly 1,300 lines.  That's a bit up there for my taste.  I do note that `summaries.go` methods are only consumed by `stdlib.go`, and that only a few `stdlib.go` methods are only consumed externally by `propagate.go`.  These feel like a good sub-package to me.

Hmmm, then I noticed the cycle of `stdlib.go::taintFromSummary -> taint.referrers`... If there's a clean lift, it still might be preferable to disentangle some of this rather than let `propagation` grow into a monolith.  But if not, then it's fine as is, especially given that half of those lines are the big summaries file.

-----

I'm surprised that the "update documentation" is (N/A) for this PR.  At a glance, though, I see that we don't explicitly document how taint propagates, instead hand-waving away "taint propagates automatically."  We really should make explicit the promises we are, and specifically aren't, making.  Then again, we should've been explicit in those promises to date and we haven't been, so I don't know that that's really an issue with _this_ PR as much as we should get to that issue in our backlog.

-----

I'll get back to this hopefully-Monday-no-later-than-Tuesday.

Gods smile upon you for the grueling task of those summaries.

PurelyApplied commented on internal/pkg/levee/testdata/src/levee_analysistest/example/tests/stdlib/funcs.go:
```diff
@@ -0,0 +1,96 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package stdlib
```
This test collection does a good job testing the positive case, but could use some negative-cases.

In that scope, many follow an `allocate, taint, sink, //want` pattern.  Adding the negative case prior to taint, i.e., `allocate sink-but-no-want, taint, sink, //want`, could be good for a wider confirmation that we aren't propagating "backwards."

mlevesquedion replied:
I felt like backwards propagation was already sufficiently covered by existing tests in e.g. `callorder`, but it can't hurt to have a few negative cases.

I will add those soon.

mlevesquedion replied:
I've tweaked a case to sink a `Writer` before it gets tainted. IMO this is sufficient since flow sensitivity is already well covered by existing cases (mostly `callorder`).

I think there is also adequate coverage for other kinds of negative cases, e.g. function returns 2 values, only one of which is tainted.

PurelyApplied replied:
LGTM

mlevesquedion commented on internal/pkg/utils/utils.go:
```diff
@@ -60,8 +60,8 @@ func DecomposeField(t types.Type, field int) (typePath, typeName, fieldName stri
 
-func unqualifiedName(v *types.Var) string {
-	packageQualifiedName := v.Type().String()
+func UnqualifiedName(t types.Type) string {
```
Doc-string: added.

As for the parameter, good observation. I've changed it back to taking a `Var`.

mlevesquedion commented on internal/pkg/propagation/stdlib.go:
```diff
@@ -0,0 +1,154 @@
+	return b.String()
+}
+
+func funcName(call *ssa.Call) string {
```
Good observation. I agree that there seems to be some redundancy here.

Actually, I think these methods deserve dedicated tests. I will add those shortly. FWIW, `utils.*` also deserve tests, but these will need to be added in a future PR.

mlevesquedion replied:
Broke out a "summary" package and added tests.

PurelyApplied replied:
Done, marking resolved.

PurelyApplied commented on internal/pkg/propagation/summaries.go:
```diff
@@ -0,0 +1,761 @@
+
+// funcSummaries contains summaries for regular functions
+// that could be called statically.
+var funcSummaries = map[string]summary{
```
Some nits throughout:

* Consider dropping the comment trailing `{`.
* Spacing is inconsistent in some interface comments, sometimes even within a single comment, e.g.
```golang
	// type context.Context interface {
	//  Err() error
	//	Value(key interface{}) interface{}
	// }
```

It smells like your IDE swapped some tabs for spaces.

mlevesquedion replied:
* What do you mean by "dropping the comment trailing `{`" ?
* Interesting. Fixed, put spaces everywhere. Side question: how did you detect that? I don't keep any visual cues for tabs/spaces so I had no idea.

PurelyApplied replied:
> What do you mean by "dropping the comment trailing {" ?
Sorry, I meant in the comment strings prior to each summary definition.  You've copy-pasted the function signature lines, which includes a trailing `{`, e.g.,

```golang
	// func Fprint(w io.Writer, a ...interface{}) (n int, err error) {
```

The trailing `{` makes me think that there should be some additional information within the comment.  The function signature is just `func Fprintf(w io.Writer, a ...interface{}) (n int, err error)` - no trailing `{`.

> how did you detect that?

Good ol' fashion eye-ballin'.  It was noticeable in the GitHub diff, even though it's not apparent in my IntelliJ view either.

mlevesquedion replied:
Ah, gotcha. Done.

mlevesquedion commented on internal/pkg/propagation/stdlib.go:
```diff
@@ -0,0 +1,154 @@
+	prop.taintFromSummary(summ, call, args, maxInstrReached, lastBlockVisited)
+}
+
+func (prop *Propagation) taintFromSummary(summ summary, call *ssa.Call, args []ssa.Value, maxInstrReached map[*ssa.BasicBlock]int, lastBlockVisited *ssa.BasicBlock) {
```
Making sure I understand: you are saying that `taintStdlibInterfaceCall` can cover both static and dynamic calls to interface methods? If so, that is my intent. (Maybe the docstrings are unclear on this point?) Specifically, given a `funcKey` such as `{"Write", "([]byte)(int,error)"}`, I want to match both of the following calls:
```
func foo(s core.Source, r io.Writer, rr myTypeThatImplementsWriter) {
  r.Write([]byte(s.Data))  // dynamic call to Writer.Write
  rr.Write([]byte(s.Data)) // static call to myTypeThatImplementsWriter.Write 
}
```

If instead you were saying that a function call might lead to `taintFromSummary` being called both through `taintStdlibCall` and `taintStdlibInterfaceCall`, unless I messed up, there should not be any overlap between the summaries for these two handlers, so that should not happen.

For the `logger` example you give, I don't think that would match any summaries, so there shouldn't be an issue?

PurelyApplied replied:
My initial comment was mostly a reminder to myself to review that the difference between `Args[0]` being the receiver or not is, in fact, exactly identified by `IsInvoke()`.  I remembered that being a stumbling block in the past, but it does look like you've got it well covered in this implementation.

Marking this chain resolved.

mlevesquedion commented on internal/pkg/levee/testdata/src/levee_analysistest/example/tests/callorder/colocation.go:
```diff
@@ -15,56 +15,31 @@
 
 import (
+	"fmt"
+	"io"
```
I actually wrote this in IntelliJ, and I'm not seeing this import re-grouping behavior. I do have a "Run goimports and save" macro bound to `Ctrl + s`, so that may be why I'm seeing the goimports behavior.

I think goimports is standard, so we should stick to that as much as possible. If I see spurious import diffs in PRs you make, I won't mind. I don't think it's reasonable to expect all contributors to open and save their files in IntelliJ before sending out a PR. :thinking:

PurelyApplied replied:
> I actually wrote this in IntelliJ, and I'm not seeing this import re-grouping behavior. I do have a "Run goimports and save" macro bound to `Ctrl + s`, so that may be why I'm seeing the goimports behavior.

Yeah, I figured it might be a personal configuration issue.  Or maybe a Community / Ultimate mismatch?  But probably just I have a box checked somewhere or an inspection turned on that you don't.  Not a big deal.

> I think goimports is standard, so we should stick to that as much as possible. If I see spurious import diffs in PRs you make, I won't mind. If we want to enforce IntelliJ formatting, I think we would need contributors to open and save each file in IntelliJ before sending out a PR, which I don't think is reasonable. 🤔

Strictly speaking, support for at least "add local imports last" is built into `goimports`, we would just need to be explicit that we want `goimports -local github.com/google/go-flow-levee,levee_analysistest`.  Easily something we could cook into our CI. 
 But it's probably not worth it in the end, so I'm happy to leave it at no harm no foul.

PurelyApplied replied:
I might.  Either way, out of scope for this PR.  Good to talk about it though.

Marking this thread resolved.

mlevesquedion commented:
Thank you for the thorough review!

Here are my replies to your high-level thoughts:
* propagation package too big? I think it's okay not to split for now.
  - As you mention, most of the lines are from summaries.
  - Logic/purpose of code in stdlib.go feels to me like it's very close to the core propagation logic
  - I may change my mind on this after I've added tests for some of the functions in `stdlib.go`.
* updated documentation (N/A)
  - Agree that we should set expectations in the docs. 
  - I think this can go in an immediately subsequent PR. High level, I'm thinking: analysis is currently intraprocedural, although the analyzer handles some standard library functions, e.g. fmt.Sprintf and any implementation of io.Write. Interprocedural propagation is in the works and expected to land sometime in 2021.

PurelyApplied commented on internal/pkg/propagation/stdlib.go:
```diff
@@ -0,0 +1,82 @@
+		prop.taint(args[i].(ssa.Node), maxInstrReached, lastBlockVisited, false)
+	}
+
+	// Taint call referrers, if there are any.
+	if call.Referrers() == nil {
+		return
+	}
```
Nit: consider moving this closer to `*call.Referrers()`.  The comment currently says "Do something" but immediately exits before moving on to something else.

mlevesquedion replied:
Ack. Reworded the comments.

PurelyApplied commented on internal/pkg/propagation/summary/testdata/src/summary_analysistest/tests/test.go:
```diff
@@ -0,0 +1,55 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package test
```
As a reminder, the `//want <string>` match as regexp.  You have the `\Q` in the one test case, but that might be useful throughout, even though I don't think we are at any real risk of matching `encoding/json2Unmarshal` or something.

However, we probably _should_ anchor the test strings with `^...$`.  This test suggests we are getting exactly `fmt.Println`, and not, for instance, `fmt.Println()`.

mlevesquedion replied:
Good point! Done.

PurelyApplied reviewed (changes requested):
I really like the refactor.  Splitting out `sumamries/api` feels good, and I'd hacked something less elegant than `For` into some local testing.  Looks good.

PurelyApplied reviewed (approved):
Looks good.

Just so it doesn't get loss in the now very long `Conversation`, I had one more non-blocking comment [here](https://github.com/google/go-flow-levee/pull/292#discussion_r606395413).

I've hacked together a before-and-after sort of comparison on [`PurelyApplied:propagation-comparisons`](https://github.com/PurelyApplied/go-flow-levee/tree/propagation-comparisons) to get a back-of-the-envelope for the sort of findings this change would have us no longer report.  Spot-checking against kubernetes, it does appear to be predominantly the `value, err := f(source)` pattern that we would now be missing, but I'm okay with that given how much this improves our signal-to-noise.  We should consider explicitly drawing attention to "you should make sure you trust your errors" when we update documentation surrounding the scope of taint propagation.
